RFC-741

TallJimbo opened the pull request:
Rendered HTML is at https://dmtn-167.lsst.io/v/DM-27147/index.html

TallJimbo committed:
First section, on raw/calib/auxilliary collections, nominally done.

TallJimbo committed:
First mostly-complete draft.

Still need to integrate RHL's late-Gen2-era doc.

TallJimbo committed:
Address initial KTL comments.

TallJimbo committed:
Fix header level.

ktlim commented on index.rst:
```diff
@@ -48,12 +48,336 @@
+   +---------------------------------------------------+-------------+-------------------------------------------------------------------------------+
+   |                   Name Pattern                    |    Type     |                                  Description                                  |
+   +===================================================+=============+===============================================================================+
+   | <instrument>/defaults                             | CHAINED     | Recommended raw, calibration, and auxilliary data for <instrument>.           |
```
```suggestion
   | <instrument>/defaults                             | CHAINED     | Recommended raw, calibration, and auxiliary data for <instrument>.           |
```
(and several other uses throughout)

MichelleGower commented on index.rst:
```diff
@@ -48,12 +48,336 @@
+There are several types of collections, which are quite different in how they how they are written, but almost identical to readers.
+A full description of the types of collections can be found in the `daf_butler documentation`_, but a brief summary here should be sufficient for the proposals in this document:
+- A ``RUN`` collection (or just "a run") is the only type of collection that is intrinsic to a dataset - a dataset is added to ``RUN`` collection when it is first added to a repository, and remains in that ``RUN`` collection until it is deleted entirely.
+- A ``CHAINED`` collection is simply an ordered list of other collections to search.  These can be nested arbitrary (but cycles are of course not permitted), and may be redefined after creation.  A single-element ``CHAINED`` collection can thus be used as a sort of symbolic link, providing a stable name for a conceptual group of datasets whose definition may change over time.
```
nested arbitrary -> nested arbitrarily?

MichelleGower commented on index.rst:
```diff
@@ -48,12 +48,336 @@
+
+.. note ::
+
+   **TODO**: get feedback from RHL and NCSA on staging directory substructure, and the workflow for ingestion.
```
I'm not sure exactly what details you're looking for here.  Currently raw images get transferred to NCSA via a couple different mechanisms (plain rsync for BOT and CCS-written images or DBB Handoff Buffer Manager service currently using bbcp).  The DBB Endpoint Buffer Manager service moves the file to a permanent read-only location in /lsstdata/offline (currently still teststand, but expected to transition to dbb subdir).  It also calls Gen2 and Gen3 ingestions (via python APIs) on each new file.  It keeps track of timings and success/failures in database tables.  Reingestion of files should, under normal circumstances, be done by "asking" the DBB BM to ingest the file again so that the database information stays valid.    We were interested in trying out the "in-place" ingestion instead of making a bunch of softlinks, but if users want to be able to manually walk filesystems and see butler directories and filenames we can continue doing the softlinks.

timj replied:
I'm actively working on the "direct" transfer mode that will put the full URI in the datastore.

timj replied:
When to run define-visits is still an interesting question in the absence of a header telling us how many observations are  in the visit.

TallJimbo replied:
I suppose it's not so much a matter of feedback as just asking for someone else to write something here; I don't know enough to be the one to do it.  I'll update the TODO note to reflect that, and we can work out later how that gets done.

And I have no problem with using in-place ingestion instead of symlinks for raws; I think I brought them up here only to emphasize that I wanted to invert the permissions/filesystems relationship: instead of `/datasets` being read-only (to devs) with symlinks there to storage on `/project`, I'd rather most of `/datasets` be read-write, with symlinks to the really important read-only-to-almost-everything storage on `/lsstdata` (and of course read-only ingest would accomplish that, too).

TallJimbo replied:
I've dropped this to-do note from the RFC (I think it's fine for the organization of `/lsstdata` to be out of scope for this document), and have proposed using in-place transfers in the text and a few new tables.

MichelleGower commented on index.rst:
```diff
@@ -48,12 +48,336 @@
+ - ``<instrument>/raw/all``: the ``RUN`` collection into which all raws for that instrument are originally ingest.
+ - ``<instrument>/raw/good/<ticket>``: a ``TAGGED`` collection containing a curated subset of all raws that do not contain problems (e.g. tracking issues, airplanes, etc.), named according to the ticket (e.g. ``DM-98765``) on which the curation work was done.
+ - ``<instrument>/raw/good``: a ``CHAINED`` collection that points to the current-best ``*/good/<ticket>`` collection for this instrument.
+
```
Apologies if I've missed it, but perhaps a reminder somewhere that rerunning with input collections with "good" in the name may not reproduce same results (as datasets could be added or removed from those collections).  Not specifically worried about the immediate users as much as looking forward to users new to LSST.

MichelleGower commented on index.rst:
```diff
@@ -48,12 +48,336 @@
+The ``pipetask`` tool will automatically take care of this if the ``--output`` option is used with or instead of the ``--output-run`` option.
+
+.. note::
+   **TODO**: does BPS provide similar functionality?
```
We would need to double check details.  BPS allows users to specify whatever output collection they want (example yaml shows "shared/ci_hsc_gen3/{timestamp}" but could easily be changed to the u pattern).  It calls pipetask with --output-run  for the --init-only execution as well as each quantum execution.  I think this came from wanting to leave the ability for production services to use other terms for uniqueness than timestamp if it wanted to (e.g., processing attempt number).

MichelleGower commented on index.rst:
```diff
@@ -48,12 +48,336 @@
+The existence of different skymap definition datasets for different coadd types (``goodSeeingCoadd_skyMap``, etc.) is something of a relic of Gen2, but one we do not plan to remove until Gen2 is fully removed.
+The new globally-unique skymap data ID names are both necessary and sufficient for Gen3, and in the future we may not store skymap definitions in datasets at all, since they must be at least partially defined in the database as well.
+
+SkyMap registration is something we expect to be rare in Gen3 - *much* more rare than running ``makeSkyMap.py`` was in Gen2 - because we almost always use one of a few standard SkyMaps, and in Gen3 a SkyMap (a combination of a ``lsst.skymap.BaseSkyMap`` class *and* its configuraiton) can only be registered once.
```
Typo configuraiton

TallJimbo committed:
Address early review comments and clean up for RFC.

TallJimbo committed:
Announce that we're moving to a single skyMap dataset type.

This is something we probably should have done long ago, but I hadn't
realized it was possible to do before Gen2 retirement.

TallJimbo committed:
More proofreading fixes and wordsmithing.

TallJimbo committed:
Don't claim to represent SciPipe leadership yet on all of this.

I doubt the rest of the science pipelines leadership has read this yet,
let alone signed off.

kfindeisen commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+This proposal is primarily concerned with long-lived, shared data repositories of the sort that will exist not just at NCSA, but at the IDF, SLAC, CCIN2P3, and other major LSST data facilities.
+
+Small repositories (typically backed by SQLite) are also expected to be common, especially for small-scale CI and local development.
+These repositories should follow the same naming patterns whenever possible, but will generally not need as many levels of indirection to guard against future changes or collections, and many of the collections defined here as ``CHAINED`` or ``TAGGED`` collections can instead be safely defined directly as ``RUN`` collections instead.
```
This guidance is a bit vague, especially since we're not used to the repository types yet. It's not obvious to new users when you can/cannot use a `RUN` collection for technical reasons (cf. my confusion on [`#dm-middleware`](https://lsstc.slack.com/archives/C2JPT1KB7/p1601312022091100) when `CALIBRATION` collections were introduced).

Do you simply mean that we can drop all the "\<ticket>" collections from the scheme, and that any `CHAINED` collections that depended on them become `RUN`/`CALIBRATION`? Might it be worth making a copy of Table 1 for this section?

TallJimbo replied:
> Do you simply mean that we can drop all the "\<ticket\>" collections from the scheme, and that any `CHAINED` collections that depended on them become `RUN`/`CALIBRATION`? Might it be worth making a copy of Table 1 for this section?

Yes, that's essentially what I meant.  I don't want to be too prescriptive on this point, but another table as an example is a good idea (but I may defer doing that until RFC adoption and the long-term home for this content is identified).

yalsayyad commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Processing initiated by DM developers that are intended primarily for personal or small-group use must start with ``u/<user>`` (e.g. ``u/jbosch``), and are strongly encouraged to start with ``u/<user>/<ticket>`` (e.g. ``u/jbosch/DM-56789``) whenever possible.
+Names and structure after this prefix are at user discretion, but we strongly recommend using a combination of ``CHAINED`` collections and ``RUN`` collections to distinguish between "inputs and outputs" collections and "output only" collections, as in :ref:`collections-shared-official-processing-outputs`.
+The ``pipetask`` tool will automatically take care of this if the ``--output`` option is used with or instead of the ``--output-run`` option.
+
```
Can we ask for clarification on this RFC PR too? These two sentences seem important but went over my head. Example?

TallJimbo replied:
If I run
```
pipetask run -b /datasets/repo --input HSC/defaults --output u/jbosch/DM-27147/test 
```
the collections I get will look like this (this option to `query-collections` is new):
```
$ butler query-collections /datasets/repo --chains=tree u/jbosch/DM-27147/test
                 Name                        Type
------------------------------------------- -----------
u/jbosch/DM-27147/test                      CHAINED
  u/jbosch/DM-27147/test/20201116T09h28m08s RUN
  HSC/defaults                              CHAINED
    HSC/raw/all                             RUN
    HSC/masks/S18A                          RUN
    refcats                                 CHAINED
      refcats/gen2                          RUN
    HSC/calib                               CHAINED
      HSC/calib/gen2/defaults               CALIBRATION
    skymaps                                 RUN
    HSC/masks                               CHAINED
      HSC/masks/S18A                        RUN
```
Indentation indicates membership in the CHAINED collection in the previous line.  The collection I specified with `--output` is the CHAINED collection at the top, and the first entry in that is a RUN collection with a timestamp-suffixed name that contains the actual processing outputs of that run.  The top-level CHAINED collection also contains `HSC/defaults`, the input collection I specified, and that recursively contains many other collections.

If I'd passed `--output-run u/jbosch/DM-27147/test/a` as well, that name would be used instead of `u/jbosch/DM-27147/test/20201116T09h28m08s`, but the content and structure would be the same.

In either case, one can initialize a butler with
```
butler = Butler("/datasets/repo", collections="u/jbosch/DM-27147/test")
```
and use it to `get` _both_ outputs and inputs, just as in Gen2.

But If one passes `--output-run` _without_ `--output`, the top-level CHAINED collection is not created, and there's no single collection that can be used to `get` outputs and inputs.

parejkoj replied:
Oh, an example like this in the text would be very useful, please!

yalsayyad commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Users are discouraged from inspecting these directories (as this will be at least quite different in the IDF or other future cloud-based datastores), and *strongly* discouraged from modifying them in any way other than via middleware tools.
+In many cases, write access will actually be prohibited (see :ref:`access-controls`).
+
+When migrations are necessary due to changes in the repository format (something that is *always* preceded by an RFC with explicit CCB approval), a new ``repo_<YYYYMMDD>`` directory and database/schema pair will be created, and files will shared via hard links until/unless the old repository is retired.
```
Do folks have something in mind re how to migrate an old database/schema? Does every schema change come with a migration script?

TallJimbo replied:
Nothing super concrete.  We have started to and are continuing to evaluate [Alembic](https://alembic.sqlalchemy.org/en/latest/) and hope to utilize as much of it as we can.  But our schema is determined from configuration as well as software versions (so _in theory_ it's a big unordered forest of versions, not a nice linear progression or even a tree), and we have a lot of non-schema content that would often need to be migrated as well, and that means we'll either be using a lot of advanced Alembic features or asking it to take care of only a subset of the whole problem.

We do have pretty good mechanisms for making it unnecessary to migrate old repositories immediately, even when code changes would create new repositories with a new schema.  And we've now got pretty good versioning embedded to detect when a migration is needed.  But for the actual migration scripts, I'll defer to @andy-slac, as he's the one who has thought about this the most.  Mostly, we're trying to avoid _any_ migrations for the next few months, to give him time to go do APDB things first and then get back to designing the migration system.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+A full description of the types of collections can be found in the `daf_butler documentation`_, but a brief summary here should be sufficient for the proposals in this document:
 
-   Feel free to delete this instructional comment.
+- A ``RUN`` collection (or just "a run") is the only type of collection that is intrinsic to a dataset - a dataset is added to ``RUN`` collection when it is first added to a repository, and remains in that ``RUN`` collection until/unless it is deleted entirely.
```
This is pretty basic, and likely entirely settled already, but why are we using `RUN` as the "base" collection? I'm assuming it's a holdover from SDSS, but I'm not sure it makes complete sense in this context. The other three names have a clearer association with their purpose, to me (although it would be interesting to know whether my "in brain from the name" image of them is actually correct!).

TallJimbo replied:
I agree that `RUN` is not a great name, though I think it's not a terrible one: it is the collection type that _always_ directly stores processing outputs, and in a long-lived repo those collections will greatly outnumber the "special" `RUN` collections that that are discussed more here (and that's where the name came from).  On the other hand, those special `RUN` collections will remain quite important, so maybe quantity isn't a great metric, but back on the original hand, users won't really interact with those special `RUN` collections directly, because we're setting up `CHAINED` collection pointers to them instead.

Overall, if I had a clearly better name in hand, I'd be tempted to see whether we could switch to it it without breaking anything (or at least without breaking repos; a lot of code would definitely have to change), but I'm not certain it's possible at this point.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+   +---------------------------------------------------+-------------+-------------------------------------------------------------------------------+
+   | <instrument>/calib/<ticket>/*                     | unspecified | Calibration production runs.                                                  |
+   +---------------------------------------------------+-------------+-------------------------------------------------------------------------------+
+   | [<instrument>/]runs/<target>/<release>/<ticket>   | CHAINED     | Public outputs of processing data <target> with <release> on <ticket>.        |
```
Here and below you transposed `/]` in `[<instrument>/]runs`

TallJimbo replied:
That was intentional; I was trying to indicate that "\<instrument\>/" is optional, and if not present, the collection should start with "run/..." not "/run/...".

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+The naming conventions for these collections are codified by the `lsst.obs.base.Instrument`_ class's ``make*Name`` methods.
+The highest-level collections are always defined as ``CHAINED`` collection "pointers" to other versioned collections.
+
+In the case of raw data, we propose three levels of collections:
```
Might be worth specifying "In the case of raw data (both science and calibration data)" just for clarity.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
 
-After consultation with other stakeholders and ultimately RFC, at least some of the content here should probably be moved to the DM Developer Guide.
+- the recommended raws for that instrument (``<instrument>/raw/good``),
+- the recommended calibrations for that instrument (``<instrument>/calib``),
```
The fact that "good" raws are in `raw/good` while "good" master calibrations are in `calib`, not `calib/good` is a bit confusing. In fact, there is no `raw` collection at all. Why the difference? `refcats` also don't have `good`, just `refcats`, so maybe we can replace `raw/good` with just `raw` too?

TallJimbo replied:
Seems like a good change to me.  I'll adopt this unless other reviewers dislike it.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+.. [#calibs-not-ingested] In Gen2, master calibration datasets *were* ingested, because the data repository in which they were produced was entirely different from the special calibration repository where they were put after certification.  Gen3 data repositories are larger, with Gen3 collections corresponding more closely to Gen2 repositories.  So certifying a master calibration in Gen3 just inv ...
+
+HSC-only auxiliary data
+^^^^^^^^^^^^^^^^^^^^^^^
```
Might be worth explicitly noting that this approach provides a useful framework for any other instrument-specific auxiliary data. I can't think of other examples, but this does mean there's an obvious place for it.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+The new globally-unique skymap data ID names are both necessary and sufficient for uniqueness in Gen3.
+
+SkyMap registration is something we expect to be rare in Gen3 - *much* more rare than running ``makeSkyMap.py`` was in Gen2 - because we almost always use one of a few standard SkyMaps, and in Gen3 a SkyMap (a combination of a ``lsst.skymap.BaseSkyMap`` class *and* its configuration) can only be registered once.
+Discrete SkyMaps, which typically cover only a small part of the sky and are *conceptually* a bit more per-user, maybe less rare, but our data model currently does not treat these any differently, and until we can identify the patterns and use cases for creating new SkyMaps (even discrete ones), we propose that any new SkyMap registration in a shared repository be preceded by an RFC.
```
"may be less rare" ?

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+The new globally-unique skymap data ID names are both necessary and sufficient for uniqueness in Gen3.
+
+SkyMap registration is something we expect to be rare in Gen3 - *much* more rare than running ``makeSkyMap.py`` was in Gen2 - because we almost always use one of a few standard SkyMaps, and in Gen3 a SkyMap (a combination of a ``lsst.skymap.BaseSkyMap`` class *and* its configuration) can only be registered once.
+Discrete SkyMaps, which typically cover only a small part of the sky and are *conceptually* a bit more per-user, maybe less rare, but our data model currently does not treat these any differently, and until we can identify the patterns and use cases for creating new SkyMaps (even discrete ones), we propose that any new SkyMap registration in a shared repository be preceded by an RFC.
```
"we propose that any new SkyMap..." - do we want to instead say "we will require any new SkyMap registration be preceded by RFC", since this is (I think?) supposed to be a guiding document.

It looks like the Access Controls section says that developers can create `skymaps`, but if we need that to be by RFC, should we restrict write access there?

TallJimbo replied:
I'm using "propose" in this document only because the RFC is still open. Some time between adopting and implementing it the language will change, possibly only when it is moved to other, more canonical documents.

I would like to give developers the ability to temporarily assume write access even for things that require an RFC; I think there's a fairly large efficiency gap between "adopt your RFC (yourself), add the skymap (yourself), mark the RFC as implemented (yourself)", and something that involves a sysadmin doing a chmod or something for you in between, and little to be gained from the latter.

parejkoj replied:
"temporarily assume" - Ok, I think that satisfies my concern. I'm more worried about "oops, I didn't mean to do that" than anything malicious, and requiring explicit adopting of a temporary role via `su` or equivalent mostly takes care of that.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Of course, the version in the collection name should differ as little as possible from the versions actually used to reduce confusion.
+
+These names should always correspond to a "public" ``CHAINED`` collection that aggregates both all ``RUN`` collections that directly hold outputs and all collections used as inputs.
+The organization of those "private" output ``RUN`` collections (if there is more than one) is completely at operator discretion, though these collections should start with the same prefix as the umbrella ``CHAINED`` collection, followed by a slash.
```
I think this is worth an example: something like `HSC/runs/RC2/w_2020_50/DM-75643/first-try` perhaps?

TallJimbo replied:
What do you want to get out of an example here?  I'm not being facetious; I'm trying to think of what an example should contain.

parejkoj replied:
I was looking for an example here because I'm having trouble figuring out what the usecase is for these "private" collections.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+The organization of those "private" output ``RUN`` collections (if there is more than one) is completely at operator discretion, though these collections should start with the same prefix as the umbrella ``CHAINED`` collection, followed by a slash.
+
+In cases where one or more private ``RUN`` collections contain datasets that should not be considered part of the final public outputs (e.g. because they are superceded by datasets in other private ``RUN`` collections), a ``TAGGED`` collection can be used to screen and aggregate these.
+That ``TAGGED`` collection would then be a direct child of the final public ``CHAINED`` collection, instead of any ``RUN`` collections it references.
```
Please include an example here: I'm having trouble following this one.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+.. note::
+
+   It is not generally possible to use a ``TAGGED`` collection as the public output collection for a processing run, because putting master calibrations (which are almost always inputs, even if indirectly) in a ``TAGGED`` collection strips them of their validity ranges and does not allow datasets from different validity ranges to coexist.
+   So even if a ``TAGGED`` collection is used, the public ``CHAINED`` collection would contain both that collection and the input ``CALIBRATION`` collection as children.
```
This paragraph suggests to me that TAGGED collections should only be allowed to be created, not written to. So, "it is generally not possible" should become "it is not possible (and enforced in code)". Which makes sense to me, if TAGGED has a similar meaning to e.g. git tags.

TallJimbo replied:
It's actually fundamentally impossible to write a dataset to a `TAGGED` collection, or any kind of collection other than a `RUN`.  They aren't really like git tags (the name comes from DESDM nomenclature for the kind of database table in which they are implemented, though it may well extend well beyond DESDM).  But they are like a set of file symlinks in the same sense that a `RUN` is like a directory, so I'd say that "writing to a TAGGED collection" is invalid in the same sense that writing a file "to a set of symlinks" is invalid.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+   **TODO**: We should resolve this uncertainty about what the last term should be in calibration production collections by the end of the RFC discussion period.
+   Gen2 used a "CalibDate", which I have always found a bit vague.
+   Some hash of input IDs seems a bit better, but it would need to be computed by external code, because we need the output collection name before we start processing (and also because the user probably wants it in some file before they launch any jobs, so they can easily look up what hashes mean).
+   Either should probably be combined with the timestamp suffixes that ``pipetask`` can automatically add to avoid clashes (especially differences to due e.g. configuration rather than inputs).
```
"to due" -> "due to"?

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+.. note::
+
+   **TODO**: While the middleware *can* use ``RUN`` collections as inputs to later CPP processing steps, it's up to the CPP team whether they want to permit that, with the alternative being to always certify between steps as a matter of policy.
```
Oh, this is a very good point. I think requiring certification between calibration-creation steps is a good idea, but that may make it more difficult to actually run calibration (e.g. can you fully vet a bias without generating darks and flats?).

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+   This leaves room for multiple curated calibration ingests to coexist, which is necessary because they will improve over time, but we don't want to assume we can remove old ones.
+   It does not provide a way to avoid duplication of curated calibration datasets that have not changed.
+
+   Calibration collections created by converting the default Gen2 calibration repo for an instrument will use ``gen2/defaults`` instead of ``<ticket>``, i.e. ``<instrument>/calib/gen2/defaults`` for the ``CALIBRATION`` collection.
```
Why `gen2/defaults` and not just `gen2`?

TallJimbo replied:
Some of our Gen2 repository sets contain multiple calib repos, and I want to leave room for more than one.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Calibration production runs intended for broad use (i.e. outputs will be at least candidates for membership in the recommended calibration collection for this instrument) should output to collections with names that start with ``<instrument>/calib/<ticket>/``.
+Those produced for experimental or development purposes should start with ``u/<user>/<ticket>/``.
+
+In either case, the ``RUN`` collections that hold output datasets directly will usually require another disambiguating term, mapping roughly to the expected validity range epoch.
```
I think this paragraph needs an example. I think you mean something like e.g. `<instrument>/calib/<ticket>/2019-11-23--2020-11-23` for a run spanning that range, but I'm not sure.

TallJimbo replied:
Agreed, but I don't want to write one until the CPP team tells me what they actually want to do here.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+   "Curated" calibration datasets that are written from a source-of-truth in an ``obs_*_data`` git repository (rather than generated directly via pipeline processing) are currently written to ``RUN`` collections with names of the form ``<instrument>/calib/curated/<calibDate>``, which are then ingested directly into an ``<instrument>/calib`` ``CALIBRATION`` collection (which clashes with our propo ...
+
+   The full workflow for curated calibrations is sufficiently unclear that it is unlikely that we will get this right in time for the first long-lived Gen3 repository.
```
It's not really part of this RFC, but this is pushing me even further into the "why do we even have curated calibrations anyway" viewpoint. We have a system for provenance and collecting calibrations described here; "curated" things may be useful as a stop-gap until this is set up, but they don't seem to serve any purpose afterwards.

timj replied:
Curated calibrations are described in DMTN-148 but there are a couple of points to make. Firstly not all calibrations are derived by DM processing the raw data -- QE curves are delivered by the camera team so have to be loaded somehow. Secondly we need a way to allow people to seed their own repos quickly without having to reprocess defects and the like and treating them as curated calibrations makes that easy without needing to export anything from the main repo.

parejkoj replied:
A way to "seed their own repos quickly" would be via butler export/import. They'll have to do that anyway to get a set of master calibs.

DMTN-148 still does not describe the workflow for how curated calibrations are created/ingested/exported-to-git, etc. I think there are still a bunch of unanswered questions there, but I don't know that any of them directly effect how the `<instrument>/calib` collections are laid out.

TallJimbo replied:
I have similar concerns about curated calibrations being underspecified, and I think it's a good point that the "seeding new repos" use case may be better served by export/import.  But I don't think we can get rid of them, and I'm resigned to not getting more clarify on how they'll work anytime soon; I think we'll have to muddle through until either things become more clear from actual usage, or some other very busy people (not it! not it!) have time to put together a more complete design.

czwa replied:
I agree that I see curated calibrations as currently used as the "repo seeds".  QE curves are the one product that I don't believe will have a calibration production workflow.  I think many of the currently curated calibrations (crosstalk and defects) will continue to want to exported in some fashion, as the construction and verification of those products is likely too expensive (in terms of computational and verification time) for most users.
I hope to get DMTN-148 updated this week to represent the "as-built" gen3 calibration/cp_pipe code.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+We will also designate two other non-repository subdirectories of ``/datasets`` for specific roles:
+
+ - ``/datasets/external`` holds files produced by other projects or surveys that may be of use to multiple users but does not fit into the LSST data model.  This includes the original versions of reference catalogs (e.g. Gaia DR2), truth catalogs (e.g. from DESC DC2), dust maps, etc.  Each subdirectory should have a descriptive README, and no files should be put in ``datasets/external`` itself.
```
It would be good to either link to our [external datasets policy](https://developer.lsst.io/services/datasets.html) or expand that info here. I'm guessing we want a similarly detailed "how to make an external dataset" doc, if we don't want to just use that one, modified for the fact that we won't have the same kind of `refcats/` subdirectory for our generated refcats.

TallJimbo replied:
My intent is for the implementation of this RFC to supersede and replace that existing dev-guide policy text.

TallJimbo replied:
FWIW, I'm recanting that part of this proposal: it's too much work to change that up (or even propose in sufficient detail how to change it up), and I'm just removing all proposals to change anything in `/datasets` right now.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+ - ``/datasets/external`` holds files produced by other projects or surveys that may be of use to multiple users but does not fit into the LSST data model.  This includes the original versions of reference catalogs (e.g. Gaia DR2), truth catalogs (e.g. from DESC DC2), dust maps, etc.  Each subdirectory should have a descriptive README, and no files should be put in ``datasets/external`` itself.
+
+ - ``/datasets/testing`` holds git LFS repositories that are used in CI and rarely change.  These are provided for convenience, and should be updated when their git master branches are; there will be no attempt to make old versions available.
```
I guess this is things like `testdata_ci_hsc` and `testdata_jointcal`? Is this for things that get `get clone` -ed for testing, or for the output gen3 repos generated from them? Is this supposed to be read-only and people can copy out of it or do a `butler import` if they need to work with the data?

TallJimbo replied:
Yes; the former; yes.

TallJimbo replied:
Also removing this from the proposal as unrelated overambition.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+ - ``/datasets/testing`` holds git LFS repositories that are used in CI and rarely change.  These are provided for convenience, and should be updated when their git master branches are; there will be no attempt to make old versions available.
+
+Finally, we propose that all raw ingestion into the shared repository be done with symbolic links or in-place (outside-root) ingests, pointing to read-only filesystems - the existing ``/lsstdata`` for Rubin Observatory data, and a new ``/external-raw`` filesystem for raw data from other instruments.
```
I'd change that from "we propose" to just "Finally, all raw ingestion..." - I see no reason that we can't make that a requirement.

TallJimbo replied:
(as above, "propose" just reflects the state of this document, not the language we'll use when it's adopted)

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+   It should not be necessary in the long term to ever use direct SQL access even for read access; the SQL schema is *not* considered a public interface - but we recognize that this may be necessary for debugging for a while.
+   This can be ensured by running::
+
+      SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;
```
Is there any way to enforce this via access controls in the database itself? I.e., if someone wants a write-allowed psql session, they have to explicitly request it?

timj replied:
In the science platform people won't have an account on the postgres database at all (DMTN-169). We are allowing developers to have accounts but expecting them not to try to subvert the system.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+This proposal specifies filesystem access controls in terms of a number of high-level "roles" that certain operators or developers may temporarily opt in to via ``su`` or special setuid tools.
+Usually these roles will be used only to create subdirectories that are owned directly by the true user.
+How to map these to users, groups, and filesystem, directory, or file-level permissions in detail is something I'd prefer to leave to the system administrators.
+All directories in ``/datasets`` will be world-readable.
```
And will all have the same groupid, with the setgid bit set?

TallJimbo replied:
That's the kind of detail I'd like to leave to the sysadmins.  I think that's how I'd do it, but I don't know if that's the only way.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+Within each ``repo_<YYYYMMDD>`` repository directory:
+
+ - Regular users will always have write access to their own ``u/<user>`` directory.
```
Do users get a `u/<user>` automatically created? Do we automatically create one when a new user gets an account?

TallJimbo replied:
I think I wrote later in the document that I'd really like the answer to be "yes".

kfindeisen replied:
Would automatic creation mean that all public repositories have a huge number of `u/` collections (`/project/` has 102 subdirectories, though not all correspond to physical persons)? That might make it hard to get one's bearings when querying collections.

ktlim replied:
I think it's easier to ensure that a `u/` directory exists and is world-writable (mode 1777 like `/tmp`), allowing users to create `u/user/` subdirectories as the Butler needs them, rather than pre-creating them for every user.  (That wouldn't be compatible with the proposal to symlink these into `/project/user`, however.)

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+ - Production operators and certain CPP team members will have access to a ``calibs`` role that can write to all ``<instrument>/calib`` directories (usually just used to create an owned subdirectory).
+
+ - Production operators and science pipelines developers who regularly ingest raw data for one or more instruments will have access to per-instrument ``<instrument>_raw`` roles that have write access to the ``<instrument>/raw`` and any auxiliary-data-collection subdirectories (e.g. ``HSC/masks``).  At present, these roles would have to be used directly whenever ingesting new raws, not just to cre ...
```
(split paragraph by sentence?)

"At present, these roles would have to be used directly whenever ingesting new raws" - why is this a bad thing?

TallJimbo replied:
> (split paragraph by sentence?)

Maybe I'm missing something obvious, but I can't figure out how to do that in bulleted lists.

> "At present, these roles would have to be used directly whenever ingesting new raws" - why is this a bad thing?

If the implementation of these access controls involves sudo-ing to some special account, I'd like to minimal how much has to be done as that account.  That's both to minimize the damage caused by accidents, and to let people use as much of their own custom environments and convenience tooling when performing nontrivial tasks.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+   If you have to do this (and not at the prompting of a middleware team member trying to help diagnose a problem), please also create a ticket explaining what you wanted to do that couldn't be done with butler tools, so we can address that feature gap.
+
+This proposal specifies filesystem access controls in terms of a number of high-level "roles" that certain operators or developers may temporarily opt in to via ``su`` or special setuid tools.
+Usually these roles will be used only to create subdirectories that are owned directly by the true user.
```
Does "true user" here mean "the user invoking the command"? Thinking about it, does it matter what e.g. the file ownership is, or what user/role runs the command? Does the registry track the user that created a dataset? If so, then we almost certainly do want the user, not the role, to run the pipeline and write the data. I don't know how that would work in actual setup.

TallJimbo replied:
> Does "true user" here mean "the user invoking the command"?

Yes.

> Does the registry track the user that created a dataset?

No; some future Science Platform version might, but right now we don't track ownership of anything in the database itself.  I'm hoping to at most _simulate_ user ownership of collections via naming conventions ("if a collection starts with `u/jbosch`, it's owned by jbosch"), and for `RUN` collections that ownership would naturally extend to the datasets within.

> Thinking about it, does it matter what e.g. the file ownership is, or what user/role runs the command?

I think it matters only in terms of who is allowed to go back and delete them later, for which the answer is usually "anyone who can assume the role that created them" (unless it's "no one, ever"), but I think we can enable either with personal ownership or role ownership depending on how we handle groups, and that's another thing I'd like to leave to the sysadmins.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+ - Production operators and science pipelines developers who regularly ingest raw data for one or more instruments will have access to per-instrument ``<instrument>_raw`` roles that have write access to the ``<instrument>/raw`` and any auxiliary-data-collection subdirectories (e.g. ``HSC/masks``).  At present, these roles would have to be used directly whenever ingesting new raws, not just to cre ...
+
+ - All production operators and science pipelines developers have access to the ``auxiliaries`` role, which provides write access to the ``refcats`` and ``skymap`` directores.  This role is used to create per-ticket subdirectories in ``refcats`` prior to starting work on ingesting a new reference catalog, and used directly to run ``butler register-skymap``.
```
(split paragraph again?)

Per my suggestion above re:skymaps, should we restrict skymap creation to a specific `skymap` role that is only granted briefly after the "make a new skymap" RFC is successful?

TallJimbo replied:
> Per my suggestion above re:skymaps, should we restrict skymap creation to a specific skymap role that is only granted briefly after the "make a new skymap" RFC is successful?

I just don't see anyone ever abusing that, in the context of a friendly, for-developers repo.

parejkoj replied:
Less a concern for abuse, and more one of accidental creation.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+   In practice, RFCs for most modifications - certainly routine ingests or calibration updates - almost never exceed our `Empowerment of DM team members`_ criteria, and asking for sysadmins who are not domain experts to do the actual work both increases friction and increases the chance something will go wrong (or, rather, the chances that if something does go wrong, it is not fixed immediately).
+
+   This proposal makes no objection to having a separate ``/project`` filesystem for files.
+   But using ``/project`` and symlinks as a way to work around restrictions on write access to appropriate subsets of ``/datasets`` is just that - a workaround - and one that makes it more difficult than it ought to be to find things.
+   If filesystem-level controls (or quotas, etc.) really are necessary even for ``/datasets`` (note that we already allow for filesystem-level controls by symlinking raws from ``/lsstdata`` and ``/external-raw``), I think this is something we can tolerate, but regular directory permission controls within ``/datasets`` would be preferable.
+   If we do use symlinks from ``/project``, we should at least ensure that a::
+
+       /datasets/repo/u/<user> -> /project/<user>/.datarepo
+
+   link (or equivalent) is automatically created with the right permissions for all users, and encourage access only via that symlink (hence the hidden directory as a target).
```
I like everything in this callout note above.

mrawls replied:
I have no objection to this, but this (automatic?) symlink between e.g. `/datasets` and `/project` needs to be clearly explained in the developer guide, because its utility and the fact that I shouldn't touch it on disk is not inherently obvious.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+This proposal is primarily concerned with long-lived, shared data repositories of the sort that will exist not just at NCSA, but at the IDF, SLAC, CCIN2P3, and other major LSST data facilities.
+
+Small repositories (typically backed by SQLite) are also expected to be common, especially for small-scale CI and local development.
+These repositories should follow the same naming patterns whenever possible, but will generally not need as many levels of indirection to guard against future changes or collections, and many of the collections defined here as ``CHAINED`` or ``TAGGED`` collections can instead be safely defined directly as ``RUN`` collections instead.
```
I see @kfindeisen asked a similar question, but I wonder whether instead of an alternate table, we can just say "you do not need to have all of the things in table 1 in your test repo"?

For my own use case, I'm thinking of the gen3 repos that get created by jointcal/fgcm from the exports file in testdata_jointcal. Currently that doesn't conform to this proposal very closely, but it could easily be made to do so, with the only real difference being that it wouldn't contain `raw/` or `u/`, and maybe not even `calib/`: only the nominal `camera` is necessary there, so there's no reason to have `calib` that I can see. Running tests could write to `<instrument>/runs/tests` or somesuch? Some guidance here might be good, so that our various CI repos don't all end up with wildly (or even worse, subtly!) different collections.

kfindeisen replied:
Avoiding "subtly different collections" is exactly why I wanted the guidance to be more concrete.

TallJimbo replied:
For the most part, I don't think one can just remove rows from table 1 without modifying the remaining rows.

I do agree that a bit more guidance is warranted, and I'm happy to add that.  But I don't think it'll resolve all questions on what to do in any particular test repo, because I'm getting the impression that pretty much each test repo differs from a standard repo in its own special way.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+"Collections" of data IDs
+-------------------------
+
+Collections that represent fields of particular interest or regularly-reprocessed test datasets are not described here, because those are conceptually more groups of data IDs than groups of datasets (e.g. not just raw exposures, but tracts on which to combine them as well).
```
"fields" here meaning "on sky observing run"?

TallJimbo replied:
Yes.  Would love to get suggestions on more precise terminology here.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+-------------------------
+
+Collections that represent fields of particular interest or regularly-reprocessed test datasets are not described here, because those are conceptually more groups of data IDs than groups of datasets (e.g. not just raw exposures, but tracts on which to combine them as well).
+As in Gen2, we will continue to record the definitions of these groups outside the data repository itself, though we may add support for in-repository storage of data IDs to Gen3 in the future.
```
What does "in-repository storage of data IDs" mean? Are you referring to a way to name a particular collection of dataIds, like `"SSP-Wide-1" = <this list of HSC visits>`? That does sound very useful, but I agree it's out of scope here.

TallJimbo replied:
Yes, essentially tables whose columns are data ID keys and whose rows contain the corresponding values.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+In the meantime, users should be aware that dataset types are *global* entities with no implicit namespacing, and hence new dataset types should be created with care.
+The ``pipetask`` tool's ``--register-dataset-types`` option is a non-default option for exactly this reason: in a long-lived repository, re-executions of the same pipeline will eventually outnumber executions of new pipelines (especially new pipelines with new datasets), and hence ``--register-dataset-types`` should rarely be needed.
+Passing it all the time as a matter of habit is an antipattern, because it makes it easy for a typo to result in long-lived, hard-to-clean-up garbage (dataset types can be removed, but only if there are no datasets of that type).
```
This seems like something that we should enforce at the code level, but I'm not sure how. Every time you run with `--register-dataset-types` a slack bot sends you a warning? ;-)

kfindeisen replied:
Objection: we do need this option on all the time in `ap_verify`, because our repositories are (supposed to be) short-lived.

TallJimbo replied:
One alternative I have been mulling is to make this a separate `pipetask` subcommand, not an option for `pipetask run`.  So you'd do
```
pipetask register-dataset-types -b repo -p mypipeline.yaml
pipetask run -b repo -p mypipeline.yaml ...
```

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+All of this fine-grained provenance is not yet implemented, however, and at present the only way to guarantee reproducibility is for all input collections to have exactly the same state they had when the original run was performed.
+The standard collections defined in this document are poorly suited for this role, however; we consider it more important for these to track the "current best" (or in the case of raws, recent observations) than it is for them remain immutable.
+Users should thus be aware that repeated processing runs using the same input collections (and everything else held constant) are *not* intended to always produce the same results (and this is a feature, not a bug).
```
+1 to everything in this provenance section: this is important, but this document only provides the foundation on which to build such provenance. I think we need to call this out loudly and often, until said reproducibility guarantees are available.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+.. _table-overview:
+
+.. table:: Overview of collection naming conventions.
```
This table is very useful, and I suspect we'll all be referring to it often to remember where things live for the first n-months of gen3 usage. My only complaint is that I had to read the rest of the document to figure out what some of the things in the table meant (e.g. the "unspecified" types), but I don't think there's any way around that. It would be nice to have links from the table name patterns to the relevant section of the doc, to make such lookups easier.

TallJimbo replied:
Links are a good idea, but something I may defer until this content is transplanted into a long-term home, as links often don't survive those kinds of transitions well.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Shared/official processing outputs
+----------------------------------
+
+Processing runs overseen by production operators should produce output collections of the form ``<instrument>/runs/<target>/<release>/<ticket>``, or ``runs/<target>/<release>/<ticket>`` in the (rare) case of processing that includes science data from multiple instruments and none of them can be considered the "primary" instrument.
```
Do we want `runs/` as the "multiple instrument" thing, or do we want some other way to denote a multi-instrument run? If I didn't know about the table, I might go looking for something in `runs/` by itself. Maybe `multi/runs` or `combined/runs` (though I'm not a huge fan of either of those)?

TallJimbo replied:
Seems reasonable.  Let's see if anyone else has any good ideas/opinions.

parejkoj reviewed (commented):
Overall thoughts: I am broadly in favour of this proposal, and in several places have suggested that "we propose" be changed to "we require" or similar, because I think it's not far off of being an actual guiding document. I do wish that we could be more strict about these conventions (see below).

Meta question: are we planning to enforce these conventions in-code? It would be useful to be able to prevent accidental "typoed" collections like e.g. `HSC/raw/goood`. Alternately, do we have ways to rename/move collections, or does one have to put them in a CHAINED collection to get the new name? I see some statement about this in the final paragraph of the Access Controls section, but I'm concerned that we may want to be a bit more strict.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Developer processing outputs
+----------------------------
+
+Processing initiated by DM developers that are intended primarily for personal or small-group use must start with ``u/<user>`` (e.g. ``u/jbosch``), and are strongly encouraged to start with ``u/<user>/<ticket>`` (e.g. ``u/jbosch/DM-56789``) whenever possible.
```
I think we want to have an explicit note in this section (and/or in "shared outputs" maybe?) that `/u/<username>/*` can CHAIN from anything, but only `/u/<username>/*` collections can be CHAINed to by other `/u/<username>/*` (possibly even requiring that the `<username>` matches?). WIthout such a requirement, if someone uses up their quota and deletes things to clean up, they could mess up other users or "official" collections. I think this should be both a written and in-code-enforced requirement.

TallJimbo replied:
> I think we want to have an explicit note in this section (and/or in "shared outputs" maybe?) that /u/<username>/* can CHAIN from anything, but only /u/<username>/* collections can be CHAINed to by other /u/<username>/* (possibly even requiring that the <username> matches?).

Agreed, as far as noting that official collections shall not depend on user collections; I'll add a note to that effect.  I do think it's important to allow users to depend on other user's collections in the shared-developer-repo context.  For the Science Platform edition, we'll need more nuance.

Right now, it's actually impossible to delete a collection if it is referenced by a CHAINED collection (including someone else's), so we have some protections in place now.  But you can still delete datasets from that collection, and still break someone else that way.  That's mostly for implementation reasons - it's just a lot harder to prevent.  But hopefully people will tend to delete entire collections rather than individual datasets, and hence it'll be hard to do.

For the curious: dataset deletion is actually one of the hardest things to get right here, because we need to maintain referential integrity between a number of database tables _and_ the content of filesystems, S3, etc. even when somebody else might be trying to write new things with the same data IDs or delete some of the same things concurrently.  A lot of our strategy involves ON DELETE CASCADE foreign key constraints, to shift a lot of that complexity to the database, and so any integrity constraint that _can't_ be captured as a real foreign key in the database is a huge pain.

mrawls commented on index.rst:
```diff
@@ -1,59 +1,367 @@
 
-   Feel free to delete this instructional comment.
+- A ``RUN`` collection (or just "a run") is the only type of collection that is intrinsic to a dataset - a dataset is added to ``RUN`` collection when it is first added to a repository, and remains in that ``RUN`` collection until/unless it is deleted entirely.
+- A ``CHAINED`` collection is simply an ordered list of other collections to search.  These can be nested arbitrarily (but cycles are of course not permitted), and may be redefined after creation.  A single-element ``CHAINED`` collection can thus be used as a sort of symbolic link, providing a stable name for a conceptual group of datasets whose definition may change over time.
```
```suggestion
- A ``CHAINED`` collection is an ordered list of other collections to search.  These can be nested arbitrarily (but cycles are of course not permitted), and may be redefined after creation.  A single-element ``CHAINED`` collection can thus be used as a sort of symbolic link, providing a stable name for a conceptual group of datasets whose definition may change over time.
```

mrawls commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+All skymaps must have a globally unique name in Gen3, which is used as part of the data ID for any dataset that is defined on tracts.
+The skymap definition datasets (i.e. ``lsst.skymap.BaseSkyMap`` subclass instances in Python) also include this globally unique name in their data IDs, and hence can also all go in a single ``skymaps`` collection.
+This is simply a ``RUN`` collection that holds skymap definition datasets directly.
```
```suggestion
This is a ``RUN`` collection that holds skymap definition datasets directly.
```

mrawls commented on index.rst:
```diff
@@ -1,59 +1,367 @@
 
-  .. figure:: /_static/filename.ext
-     :name: fig-label
+Collections
```
This document begins with terminology definitions, and not with its scope. I had to spend a lot of time digesting everything before I learned it is a set of requirements for big, long-term, shared data repositories that may live at any one of a number of locations and may or may not use a POSIX filesystem.

Throughout, the jargon is heavy, and examples would go a long way to add clarity. For example, each pattern in Table 1 could have 1-2 examples, and then immediately afterwards you could briefly define run, tagged, chained, etc.

TallJimbo replied:
I hope we'll get to that level of documentation eventually, and I certainly understand that it's hard to read a policy proposal like this without the background needed to make sense of it (especially when documentation for that background doesn't make sense), but I just don't plan to evolve this into a user guide or Gen3 intoduction.  The daf_butler package docs and various scattered live-tutorial materials (which need to be integrated into the pipeline docs, of course) are already better for that, even if "better" != "good".

On the scope question: that was exactly what I intended to cover in the abstract (i.e. the paragraph above this heading).  Having gotten to the end, is there something specific I should have included in the abstract that would have helped, or some other reordering of this section with the others that would have worked better?  I did terminology right after the abstract specifically because I was worried about leading with undefined jargon, and felt it was appropriate to cover a little bit of that instead of just linking back to daf_butler docs.  But perhaps that just gives the wrong impression about the purpose of the document.

mrawls replied:
I will concede that I skimmed over the second half of the abstract :) I saw this

>This document will be - at least at first - a proposal for how to organize data repositories in detail, focusing on collection naming conventions, filesystem locations, and developer workflows.

And did not fully appreciate the subsequent clarification that everything after assumes the "one big shared repository" model.

I still think the heavy focus on terminology is a lot to wade through, and the first part of Section 1 would be best used introducing the how and the why of the single shared repository model (leading naturally into collections within it).

TallJimbo replied:
I've added another paragraph to the very first section, explaining why I'm advocating one big shared repo.

mrawls commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+Shared/official processing outputs
+----------------------------------
+
+Processing runs overseen by production operators should produce output collections of the form ``<instrument>/runs/<target>/<release>/<ticket>``, or ``runs/<target>/<release>/<ticket>`` in the (rare) case of processing that includes science data from multiple instruments and none of them can be considered the "primary" instrument.
```
I don't know how to tell if/when I am a production operator doing an official processing run (`<instrument>/runs`) or a developer doing processing for myself or a small group (`u/<user>`).

TallJimbo replied:
I think you and I (and other people at generalized Princeton/UW) are not really ever "production operators", at least in the construction Project - that's the title I'm using for (e.g.) the roles currently played by Monika and Eric at NCSA.  Beyond that, I'd defer to @MichelleGower on this question, as I picked up the terminology from her and mostly inferred its meaning (and the notion that there is a clear separation) from context in various conversations, rather than having a clear definition of it in my own head.

mrawls replied:
That is interesting and not the answer I expected! Further clarification on this would be welcome.

MichelleGower replied:
My assumption is that there's a production operator role.  Members of that role know when they are doing production runs vs doing personal science runs.  Monika and Eric are currently the folks at NCSA in that role (e.g., they are responsible for running the HSC RC2 pipelines).  (pre-) Operations will determine who those folks are for the data previews and data releases.  I'm pretty sure I haven't answered the question, but I'm not quite sure what the question is.  What was the answer you expected?

mrawls replied:
That's closer to what I expected :) I guess I thought some project developers would share responsibility with data facility folks to keep new datasets ingested and/or processed. Thank you both, I recognize this is still being defined and precursor datasets are a bit of a gray area.

mrawls commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+Processing runs overseen by production operators should produce output collections of the form ``<instrument>/runs/<target>/<release>/<ticket>``, or ``runs/<target>/<release>/<ticket>`` in the (rare) case of processing that includes science data from multiple instruments and none of them can be considered the "primary" instrument.
+``<target>`` is a human-meaningful name for the set of data IDs being processed, and ``<release>`` is some kind of DM software release version, so examples of complete processing-output collection names might include ``HSC/runs/RC2/w_2020_50/DM-75643`` or ``DECam/runs/HiTS-2015/d_2021_90/DM-80000``.
+These versions are intended to make it easy for users to browse collections and understand what is in them at a glance; formal provenance for software versions actually used in the processing will be automatically stored in the data repository itself.
```
I appreciate this sentiment, but worry it may be cumbersome. Shouldn't the Butler be able to tell me pretty quickly what release version was used? What if I want to use a naming scheme other than a ticket because some work spans multiple tickets or is for an un-ticketed science project?

TallJimbo replied:
This naming scheme is just a slight adjustment to what we've used in Gen2 for HSC RC2 processing runs, which I think has worked pretty well in practice for the DRP people consuming those runs, though I'd be interested to hear from the operators whether there's room for improvement from their perspective.

As to using other naming schemes for other projects, I think that gets back to the "who is an operator" question; I think the cases you mention just aren't the "overseen by operators" ones, and they'd hence go in `u/<user>` (and could use whatever naming convention you'd like after that prefix).

Butler can tell you which version was used to produce any specific dataset with great precision, but it's harder for it to reason about approximate/summary versions for a larger processing campaign, especially if, say, there was a patch applied to one package and thus a slightly different version used for some of the processing.  But I was actually thinking more about the converse: I want to to _find_ the processing run for w_2020_51, and it's there that I think including the release in the collection name (and doing it before the ticket number!) is particularly helpful.  It's possible that thinking is a holdover from Gen2, in which you'd poke around the filesystem by descending into directories, and the ordering of subdirectory terms is very important.  But I don't think it's harmful for Gen3, where poking around probably involves glob expressions on the whole collection name instead.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+In the case of raw data, we propose three levels of collections:
+
+ - ``<instrument>/raw/all``: the ``RUN`` collection into which all raws for that instrument are originally ingested.
```
Should we allow users to ingest raws into their `u/<user>` collections, and if so should we specify the convention for naming userspace raw collections; possibly `/u/<user>/<instrument>/raw`? If we recommend against that (can we even prevent it?), what is the process for getting new data ingested if you are not in the `<instrument>_raw` role group?

TallJimbo replied:
I do strongly recommend against that, but it's hard to prevent right now; if I can find a way to make it impossible without assuming a non-user role, I will.  The big problem is that raw ingestion does not just ingest datasets - it also inserts `exposure` dimension rows, and if some user does that _before_ it is done officially, and something goes wrong, it'll be hard to clean things up in order to let the official ingestion run proceed.

> what is the process for getting new data ingested if you are not in the <instrument>_raw role group?

I think this'll be a bit different for each instrument, but to first order, I think it's basically, "ask your friendly neighborhood member of that group".

parejkoj replied:
@mrawls has strong opinions here, regarding science users who want to process e.g. last night's DECam data right away, when we don't have an official ingest for it. That may be something that we explicitly do not support on this system as it is (and as you say, we may want to make it impossible), but it is something we'll want to think about for the RSP.

parejkoj commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+The current Gen3 registry architecture does not allow any fine-grained access control in the repository database; we instead rely on "friendly users" being careful and respectful of this shared space.
+
+At the same time, we will use filesystem access controls to protect shared and per-user files, and we plan to implement some checks in the Butler client itself to make it at least extremely difficult to *accidentally* cause problems.
```
How should we handle users who want to make their user space `collections` unreadable to anyone else, or grant permission to specific people to access some but not all of their `collections`?

timj replied:
We are not going to support that for our developers at NCSA. Everyone will be able to see everyone else's collections.

In the science platform we will have a client/server registry where the server will know your user ID and your group membership.

mrawls commented on README.rst:
```diff
@@ -14,7 +14,7 @@ Policies and Conventions for Organizing Gen3 Data Repositories
 ========
 
-While the Gen3 Butler provides some intrinsic structure to its data repositories, considerably more is left to convention (often encoded in higher-level packages, like obs_base).  This document will be - at least at first - a proposal from Science Pipelines leadership on how to organize data repositories in detail, focusing on collection naming conventions, filesystem locations, and developer wor ...
+While the Gen3 Butler provides some intrinsic structure to its data repositories, considerably more is left to convention (often encoded in higher-level packages, like obs_base).  This document will be - at least at first - a proposal for how to organize data repositories in detail, focusing on collection naming conventions, filesystem locations, and developer workflows.  The immediate focus will ...
```
This document needs to say upfront that it assumes a model of one butler repository to rule them all (`/datasets/repo`) with lots of options for collections. Is there ever any guidance for the situation where users or developers need or choose to create an entirely different repository? Or for the situation where users or developers want to ingest new data into the "one repo to rule them all" but may not have the necessary permissions?

TallJimbo replied:
:+1:; I'll add this to the abstract:

> A core assumption is that there will be _one_ large shared data repository for all instruments at NCSA (and each other major facility), for "friendly" use by the DM team.
Smaller custom data repositories may also exist (especially for CI), but we hope to ensure that nearly all development work can be performed without the need for developers to ever create their own personal repositories.
One or more large shared repositories for science users are also expected to exist, but are explicitly beyond the scope of this proposal.

(I think that may also help address your earlier comment about starting with terminology instead of scope.)

mrawls commented on index.rst:
```diff
@@ -1,59 +1,367 @@
+
+.. note ::
+
+   **TODO**: This document should probably say more about staging locations for (at least) raws that haven't been ingested yet, for both Rubin Observatory instruments and precursor instruments (which I imagine might be quite different in this respect).  I'd like to at least be confident that the filesystem locations and permissions I'm proposing are consistent with how data arrives and gets inges ...
+
```
This leaves me wondering where raws should land on the filesystem to ensure they aren't accidentally blown away after ingestion. I also really want to know what butler hoops I need to jump through to get a cold hard FITS file (at the end of some processing pipeline, from a particular collection) in my science-hand.

timj replied:
Butler doesn't need to have the raw files in the datastore at all since it can refer to the primary location in `/datasets`. This would make it impossible to delete them even if you tried. If we are using softlinks instead we would want to set permissions on that part of datastore to make it impossible for users to delete anything.

TallJimbo replied:
> I also really want to know what butler hoops I need to jump through to get a cold hard FITS file (at the end of some processing pipeline, from a particular collection) in my science-hand.

Butler will at least be able to guarantee that it can get you a URI to that FITS file, assuming it exists as a [single] FITS file, and as a DM dev you'll probably know what to do with that URI.

But if you drop some of those assumptions, this becomes more of a DAX / Science Platform question, and while I'd expect the butler to be involved in the implementation of giving science users FITS files for things that maybe weren't really stored that way, it feels a little scope-creepy (to me, at least) to make it the front-end for that.

mrawls replied:
My main concern here relates to @parejkoj's example of a science user wanting to process e.g. yesterday's DECam observations. As I understand it, if I have a new pile of last week's DECam images, I have two options: (1) ask around informally until I find somebody with permission and know-how to add things to `/datasets` and the "one giant repository" for me or (2) start my own Butler repository completely independent of everything else, which feels antithetical to the spirit of this entire document, especially if my amazing images turn out to be useful to the Project.

What I am looking for is a section of this document which improves markedly on the current state of (1), and perhaps gives guidelines for when (2) is and is not acceptable.

MichelleGower replied:
If ingestion of DECam raw files is interesting to the project in general, ask @mbutlerncsa and @rgruendl about setting up automatic (transfer and) ingestion of those raw files.

mrawls reviewed (commented):
I don't have objections to the proposed conventions, but there are lots of opportunities to improve clarity and explain how things work to a broader audience. I spent several hours today reading this, doing my best to comprehend it, and sorting out which parts of it apply to me when I don my developer hat vs. my science user hat. I strongly echo other comments requesting more examples throughout. Significant work remains to create more user-facing documentation about navigating and utilizing shared repositories and collections.

airnandez commented:
I would like to comment on a topic related to the section [2 Filesystem locations](https://github.com/lsst-dm/dmtn-167/blob/tickets/DM-27147/index.rst#2filesystem-locations) even if it is not explicitly mentioned there.

From my ingestion tests of raw image data of HSC RC2 using **w_2020_47** I noticed that the resulting (POSIX) Butler repo is organized as:

```
$ tree -L 6 .
.
├── butler.yaml
└── HSC
    └── raw
        └── all
            └── raw
                ├── g
                │   └── HSC-G
                ├── i
                │   └── HSC-I
                ├── r
                │   └── HSC-R
                ├── y
                │   └── HSC-Y
                └── z
                    └── HSC-Z
```
(I did not modify the templates so I guess this is the default organization)

Within a leaf filter directory (e.g. `HSC-G` in the example above) the raw files are named with the pattern like

```
raw_g_HSC-G_34486_98_HSC_HSC_raw_all.fits
```

If I extrapolate from this observation, each leaf filter directory would contain all the FITS files for all the visits for that particular filter. If this is the envisioned organization of raw data for the LSSTCam instrument, I would like to express my concern regarding the number of files we would end up with in a single (filter) directory.

Assuming:

* 2000 visits per night
* 300 nights per year
* 10 years
* 6 filters
* 189 sensors

we would end up with 1M visits per filter (on average) and 189M files per filter directory. I suspect this situation is not desirable for any POSIX file system and may not be neither for any contemporary object store.

I wonder if we should introduce additional levels in the namespace, for instance, acquisition date or visit ranges, or something that could help us ending up with a manageable number of objects in each level of the hierarchy.

In addition, I would propose that we adopt simple conventions in the naming of those files, for instance:

* use `098` for the identifier of the sensor,
* use `00034486` for the identifier of the visit,

Although not important for storage systems, I think those kind of conventions may be useful for humans, for instance when retrieving the list of all files in a given visit or the list of visits in a given range, etc.

(Apologies if this comment doesn't belong to this RFC. I would be happy to move it to the appropriate place if that is the case.)

timj commented:
@airnandez thank you for thinking about the file templates. We actually define a single template for all data files: https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/configs/datastores/fileDatastore.yaml#L8 but we can definitely define a different template for raw.

Zero-padding the detector is possible but the amount of padding is instrument specific. We can do instrument-specific raw templates if desired.

For the exposure ID people might prefer we use the obsid string, especially for LSSTCam and LATISS. We could use that and a padded detector easily enough.

I take your point on using a date-stamped directory.  Currently you can say `exposure.timespan_begin` in the file template but there is no way to make that use a YYYY-MM-DD format (it would return the full ISO date string). It's fairly simple to add a new format code that will allow that though.

TallJimbo commented:
General reply to @parejkoj 's general comments:

> Meta question: are we planning to enforce these conventions in-code?

Yes, but it may not be until after the first long-lived repo is stood up, and the enforcement will not be complete for a long while (maybe never for repos that only allow internal devleoper access).  The Science Platform version of butler will be quite different in this respect.

> It would be useful to be able to prevent accidental "typoed" collections like e.g. HSC/raw/goood.

Typos in _input_ collections should always generate errors, and I think we've _mostly_ done a good job of making those errors happen early enough that no modifications are made to the repo, though this varies from command to command.

> Alternately, do we have ways to rename/move collections, or does one have to put them in a CHAINED collection to get the new name? I see some statement about this in the final paragraph of the Access Controls section, but I'm concerned that we may want to be a bit more strict.

We do not have a way to rename or move collections, aside from manual database surgery.  It may be possible to add that in some cases, but I'm hoping to make it as rare as possible.

TallJimbo committed:
Simple corrections and improvements from John Parejko's review.

TallJimbo committed:
Simple corrections and improvements from Meredith Rawl's review.

czwa commented on index.rst:
```diff
@@ -1,59 +1,374 @@
+
+.. note::
+
+   **TODO**: While the middleware *can* use ``RUN`` collections as inputs to later CPP processing steps, it's up to the CPP team whether they want to permit that, with the alternative being to always certify between steps as a matter of policy.
```
I had not understood that a calibration in a RUN collection could be used directly without having been certified into a CALIBRATION collection.  This may allow a more efficient calibration production when success is expected, as all certification can be done after all calibrations are finished (and using a single processing pipeline).  I think this should likely be allowed, but not be the default production method until more experience with calibration production exists.

czwa commented on index.rst:
```diff
@@ -1,59 +1,374 @@
+Calibration production runs intended for broad use (i.e. outputs will be at least candidates for membership in the recommended calibration collection for this instrument) should output to collections with names that start with ``<instrument>/calib/<ticket>/``.
+Those produced for experimental or development purposes should start with ``u/<user>/<ticket>/``.
+
+In either case, the ``RUN`` collections that hold output datasets directly will usually require another disambiguating term, mapping roughly to the expected validity range epoch.
```
I believe that the disambiguating term used for the RUN collections should simply be an estimated validStart value.  New calibrations (with updated validStart/validEnd values) will be necessary to track the time evolution of the camera-telescope system.  With the verification system currently in development, we should have metrics illustrating the quality of the currently certified calibrations, allowing us to track this evolution.  This means the new validStart date should not be completely unknown prior to calibration production.  This naming convention avoids the need to calculate the output collection from the input data (a problem with hashing), and should provide a human readable collection name.
I don't think a full range is helpful here, as calibrations are expected to be valid until replaced.  When generating the new calibration, we will likely always know the start date we want to use, but rarely the end date.

TallJimbo replied:
The only qualm I have about using `validStart` as the final disambiguating term is that (because this term is essentially frozen into a dataset's identity forever), it is conceivable that a dataset could have a different actual `validStart` in one or more CALIBRATION collection, and hence users will have to be careful about identical or near-identical values actually representing different things in different contexts, and especially queries for that datasets that show a now-misleading RUN collection name.  I don't think that's a sufficient reason to drop the idea, but I wanted to make sure you were aware of the wrinkle.

TallJimbo replied:
I've included this in the text and removed the TODO.

czwa commented on index.rst:
```diff
@@ -1,59 +1,374 @@
+   Some hash of input IDs seems a bit better, but it would need to be computed by external code, because we need the output collection name before we start processing (and also because the user probably wants it in some file before they launch any jobs, so they can easily look up what hashes mean).
+   Either should probably be combined with the timestamp suffixes that ``pipetask`` can automatically add to avoid clashes (especially differences  due to e.g. configuration rather than inputs).
+
+As noted in :ref:`collections-per-instrument`, certified calibration products intended for broad use should go in ``CALIBRATION`` collections named *just* ``<instrument>/calib/<ticket>``.
```
Following the text, the output products from each of a ticket's calibration production RUN collections would then be certified into that ticket's final CALIBRATION collection.  Will the final \<instrument>/calib be a chained collection of these ticketed collection?  If so, then this seems like observed changes in the telescope only require new tickets to generate new calibrations after the estimated validStart date, which can then join the chained \<instrument>/calib.  I believe this system also works with algorithmic improvements: using the existing validStart values, we can regenerate updated calibrations for each date range.

TallJimbo replied:
I would prefer to make `<instrument>/calib` a CHAINED collection that points to just one CALIBRATION collection that contains a version of everything, e.g. by having a new ticket's CALIBRATION collection copy certifications (but not datasets) from whatever was considered the previous best CALIBRATION collection.  That will rigorously enforce the constraint on one dataset per data ID and dataset type at any point in time, and it should make it a lot easier to generate fast queries for calibration lookups in `QuantumGraph` generation.

czwa commented on index.rst:
```diff
@@ -1,59 +1,374 @@
+
+.. note::
+
+   "Curated" calibration datasets that are written from a source-of-truth in an ``obs_*_data`` git repository (rather than generated directly via pipeline processing) are currently written to ``RUN`` collections with names of the form ``<instrument>/calib/curated/<calibDate>``, which are then ingested directly into an ``<instrument>/calib`` ``CALIBRATION`` collection (which clashes with our propo ...
```
Using the structure mentioned above matches what the curated calibrations will also do: updated curated calibrations can be ingested to a tree of ticketed collections, split by the same date value that currently exists in the obs_*_data packages.  These calibrations can then be certified into same ticketed location that they would appear had they been generated directly on that ticket.
I don't think we want to call the disambiguating term either the CalibDate or the validStart.  I'm been calling this an estimatedDate while thinking this through, as it's a guess as to what the validStart date will be.  In the case of curated calibrations, this guess is assumed to be perfectly accurate, as the curation process has done the additional validation checks to determine what the validStart will be.

TallJimbo committed:
Update to new paths, new DC2 approach, and simpler access controls.

TallJimbo committed:
Updates from proofing and PR comments.

TallJimbo committed:
Resolve some calibration TODO notes.

TallJimbo committed:
Change disclaimer note to reflect RFC-741 adoption.
